src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+        Ok(())
+    }
+
+    fn build(self) -> (Option<Document>, Option<Document>) {
+        let (mut upsert_document, mut delete_document) = (None, None);
+        if !self.upserts.is_empty() {
+            upsert_document = Some(doc! {
+                "update": self.coll.clone(),
+                "ordered": true,
+                "updates": self.upserts,
+            });
+        }
+        if !self.deletes.is_empty() {
+            let deletes = self
+                .deletes
+                .into_values()
+                .map(|pk| {
+                    bson!({
+                        "q": pk,
+                        "limit": 1,
+                    })
+                })
+                .collect::<Array>();
+
+            delete_document = Some(doc! {
+                "delete": self.coll,
+                "ordered": true,
+                "deletes": deletes,
+            });
+        }
+        (upsert_document, delete_document)
+    }
+}
+
+type MongodbNamespace = (String, String);
+
+// In the future, we may build the payload into RawBSON to gain a better performance.
+// The current API (mongodb-2.8.2) lacks the support of writing RawBSON.
+struct MongodbPayloadWriter {
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    default_namespace: Namespace,
+    coll_name_field_index: Option<usize>,
+    ignore_coll_name_field: bool,
+    is_append_only: bool,
+    client: Client,
+    buffered_entries: usize,
+    max_entries: usize,
+    // TODO switching to bulk write API when mongodb driver supports it
+    insert_builder: Option<HashMap<MongodbNamespace, InsertCommandBuilder>>,
+    upsert_builder: Option<HashMap<MongodbNamespace, UpsertCommandBuilder>>,
+}
+
+impl MongodbPayloadWriter {
+    fn new(
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        default_namespace: Namespace,
+        coll_name_field_index: Option<usize>,
+        ignore_coll_name_field: bool,
+        is_append_only: bool,
+        client: Client,
+        max_entries: usize,
+    ) -> Self {
+        Self {
+            schema,
+            pk_indices,
+            default_namespace,
+            coll_name_field_index,
+            ignore_coll_name_field,
+            is_append_only,
+            client,
+            buffered_entries: 0,
+            max_entries,
+            insert_builder: if is_append_only {
+                Some(HashMap::new())
+            } else {
+                None
+            },
+            upsert_builder: if is_append_only {
+                None
+            } else {
+                Some(HashMap::new())
+            },
+        }
+    }
+
+    #[inline(always)]
+    fn document_from_row_ref<'a>(&'a mut self, row: RowRef<'a>) -> Document {
+        // Why there is no Document::with_capacity in bson crate?
+        self.schema
+            .fields()
+            .iter()
+            .zip_eq_debug(row.iter())
+            .enumerate()
+            .filter_map(|(index, (field, datum))| {
```

Comment:
It seems that the index information is fixed for any input. If so, we can pre-calculated the index information to avoid comparing index for each row.